File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/I._Сложное_поле_с_цветочками/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(NM)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    grid: list[str] = []
    for _ in range(n):
        row: str = sys.stdin.readline().rstrip()
        grid.append(row)
    grid.reverse()
    dp: list[list[int]] = [[0] * m for _ in range(n)]
    for row in range(n):
        for col in range(m):
            dp[row][col] = int(grid[row][col]) + max((dp[row - 1][col] if row else 0), (dp[row][col - 1] if col else 0))
    sys.stdout.write(f'{dp[n - 1][m - 1]}\n')
    stacktrace: list[str] = []
    row, col = n - 1, m - 1
    while row or col:
        right: int = dp[row][col - 1] if col else float('-inf')
        up: int = dp[row - 1][col] if row else float('-inf')
        if right >= up:
            stacktrace.append('R')
            col -= 1
        else:
            stacktrace.append('U')
            row -= 1
    stacktrace.reverse()
    sys.stdout.write(''.join(stacktrace) + '\n')



if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/I._Сложное_поле_с_цветочками/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>


void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(NM)
    */
    int n, m;
    std::cin >> n >> m;
    std::vector<std::string> grid;
    for (int i = 0; i < n; ++i) {
        std::string row;
        std::cin >> row;
        grid.push_back(row);
    }
    std::reverse(grid.begin(), grid.end());
    std::vector<std::vector<int>> dp (n, std::vector<int>(m, 0));
    for (int row = 0; row < n; ++row) {
        for (int col = 0; col < m; ++col) {
            dp[row][col] = (grid[row][col] - '0') + std::max((row? dp[row - 1][col] : 0), (col? dp[row][col - 1] : 0));
        }
    }
    std::cout << dp[n - 1][m - 1] << "\n";
    std::string stacktrace = "";
    int row = n - 1, col = m - 1;
    while (row || col) {
        int right = (col? dp[row][col - 1] : -1);
        int up = (row? dp[row - 1][col] : -1);
        if (right >= up) {
            stacktrace.push_back('R');
            --col;
        } else {
            stacktrace.push_back('U');
            --row;
        }
    }
    std::reverse(stacktrace.begin(), stacktrace.end());
    std::cout << stacktrace << "\n";
}

int main() {
    solution();
}